### PR TITLE
FEATURE: RAIL-2044 Add 'saveAs' and 'cascadingDelete' functions for d…

### PR DESCRIPTION
**Pull request by Tulean**
…ashboards

Added 'saveAs' functions that clones a specific dashboard with either same or new object references.
Function 'cascadingDelete' deletes dashboard and its referenced objects.

---

# PR checklist

- [ ] Change was tested using dev-release in Analytical Designer and Dashboards (if applicable).
- [ ] Change is described in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [ ] Migration guide (for major update) is written to [CHANGELOG.md](../blob/master/CHANGELOG.md).

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/55/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/485/ : FAILURE in 2m 11s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/56/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/486/ : FAILURE in 46s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/57/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/487/ : FAILURE in 1m 21s

File: src/metadataExt.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
+// (C) 2020 GoodData Corporation
+import { MetadataModule } from "./metadata";
+import { UserModule } from "./user";
+import { XhrModule } from "./xhr";
+import { IGetObjectUsingOptions } from "./interfaces";
+import cloneDeepWith from "lodash/cloneDeepWith";
+
+export interface ICopyDashboardOptions {
```
Comment (lupko) [inline src/metadataExt.ts]: please add documentation to these options; on interface & then on props as well

File: src/metadataExt.ts [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+                },
+            },
+        });
+    }
+
+    private getDashboardName(originalName: string, newName?: string) {
+        if (newName !== undefined) {
+            return newName;
+        }
+        return `copyOf(${originalName})`;
```
Comment (lupko) [inline src/metadataExt.ts]: the default name should be less 'techy'.  i propose minor reword 'Copy of ${originalName}'

**Comment by lupko** [comment]
@Tulean the PR contains a bunch of file changes that do not seem to be related to your PR; they contain only formatting changes + copyright change.

please either remove these files or have them in a separate TRIVIAL: Improve formatting commit or something like that

File: src/metadataExt.ts [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            },
+        });
+
+        const dashboardID: string = analyticalDashboard.meta.uri.split("/")[5];
+        const types = ["visualizationObject", "kpi", "visualizationWidget", "filterContext"];
+        const objectsUsingDashboard: any[] = await this.getObjectsUsingDashboard(projectID, dashboardUri, {
+            types,
+        });
+
+        if (options.keepReferences) {
```
Comment (lupko) [inline src/metadataExt.ts]: this is not the desired behavior. we want this:

- Save As always copies Dashboard + Widgets + Filter Context
- SaveAs options can be used to control whether: KPIs referenced by Widgets should be copied, Insights referenced by Widgets should be copied

Comment (Tulean) [reply src/metadataExt.ts]: Got it. Will be working on it.

File: src/metadataExt.ts [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+            analyticalDashboard: {
+                ...dashboard.analyticalDashboard,
+                meta: {
+                    ...dashboard.analyticalDashboard.meta,
+                    title: this.getDashboardName(dashboard.analyticalDashboard.meta.title, options.name),
+                },
+            },
+        });
+
+        const dashboardID: string = analyticalDashboard.meta.uri.split("/")[5];
```
Comment (lupko) [inline src/metadataExt.ts]: I think there might already by a function in gd-js to obtain object ID from URI. please look around, it could be somewhere around utils; if the function does not exist then I suggest to create one and use it in all places where we need to get object ID from URL.

Comment (lupko) [reply src/metadataExt.ts]: ideally do something like getObjectIdForObj(obj).. and the getObjectId() looks whether obj.meta.uri is available and if so do whatever

File: src/metadataExt.ts [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+     * @param {string} projectId Project identifier
+     * @param {string} dashboardUri Uri of a dashboard to be deleted
+     */
+
+    public async cascadingDelete(projectID: string, dashboardUri: string): Promise<any> {
+        const objects = await this.metadataModule.getObjectUsing(projectID, dashboardUri);
+        const currentUserInfo = await (await this.userModule.getAccountInfo()).profileUri;
+
+        const objectsToBeDeleted = objects
+            .filter((object: any) => {
```
Comment (lupko) [inline src/metadataExt.ts]: you can use shorten this to filter((object: any) => object.author === currentUserInfo)

please apply this sugar to all simple lambdas which just do return <something>

File: src/metadataExt.ts [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        const { analyticalDashboard } = await this.metadataModule.createObject(projectID, {
+            ...dashboard,
+            analyticalDashboard: {
+                ...dashboard.analyticalDashboard,
+                meta: {
+                    ...dashboard.analyticalDashboard.meta,
+                    title: this.getDashboardName(dashboard.analyticalDashboard.meta.title, options.name),
+                },
+            },
+        });
```
Comment (lupko) [inline src/metadataExt.ts]: this is unfortunate :( new dashboard object should be the last thing the saveAs does - after everything else has succeeded. the content of the new dashboard object can be kept and updated in memory and stored just once after all is done

File: src/metadataExt.ts [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+        const { uri }: { uri: string } = await this.metadataModule.updateObject(projectID, dashboardID, {
+            analyticalDashboard: {
+                ...analyticalDashboard,
+                content: {
+                    ...analyticalDashboard.content,
+                    layout: { ...analyticalDashboard.content.layout, fluidLayout: clonedLayout },
+                    widgets: await replaceWidgetsRef,
+                },
+            },
+        });
```
Comment (lupko) [inline src/metadataExt.ts]: this will not be needed once saveAs creates new dashboard object at the very end of processing. until then you can pass around the copy of analytical dashboard, keep updating the links in it; returning the updated object after each function

File: src/metadataExt.ts [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+        const { uri }: { uri: string } = await this.metadataModule.updateObject(projectId, dashboardID, {
+            analyticalDashboard: {
+                ...analyticalDashboard,
+                content: {
+                    ...analyticalDashboard.content,
+                    filterContext: filterContext.meta.uri,
+                },
+            },
+        });
```
Comment (lupko) [inline src/metadataExt.ts]: same as before, this should go away if you create new instance of dashboard at the very end of saveAs processing. until then you can pass around and work with dashboard object in memory

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/65/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/495/ : SUCCESS in 2m 48s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/66/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/496/ : SUCCESS in 54s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/67/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/497/ : SUCCESS in 1m 49s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/68/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/498/ : SUCCESS in 58s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/69/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/499/ : SUCCESS in 51s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/70/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/501/ : SUCCESS in 4m 54s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/71/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/502/ : SUCCESS in 2m 09s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/72/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/503/ : SUCCESS in 1m 39s

**Comment by lupko** [comment]
we did code review offline & had several iterations over the code. at the moment this looks good. we will be continuing on the ticket and refine few things: add typings, do further refactor for testability & improve error handling.

another PR will follow

**Comment by lupko** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-merge-compliance-check/116/ : SUCCESS in 43s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/504/ : SUCCESS in 3m 56s
